[Cache] Add `PdoTagAwareAdapter`

Toflar opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | 
| License       | MIT

I noticed that unfortunately, the  `PdoAdapter` in the cache component does not natively support cache tagging. This forces you to use it in combination with the general `TagAwareAdapter` which has a few downsides when it comes to performance. So I figured, why not try to add native support.

## How it works

Tags live in their own table, one `(item_id, item_tag)` row per tag, with an index on the tag column:

```sql
DELETE FROM cache_items WHERE item_id IN (SELECT item_id FROM cache_tags WHERE item_tag IN (:tags))
```

So saving, invalidating and pruning are single relational statements, where wrapping `PdoAdapter` in `TagAwareAdapter` costs extra round trips per operation.

The PDO plumbing shared by both adapters moves from `PdoAdapter` into a new internal `PdoTrait`; `PdoAdapter` itself is behaviorally unchanged. `AbstractTagAwareAdapter::TAGS_PREFIX` becomes `protected` and is read with `static::`, so this adapter can redeclare it empty: the prefix exists to separate tags from items in a shared key space, which a dedicated table does not have. The constant's docblock states that it takes part in the persisted keys, so redeclaring it on an existing pool orphans the tag entries already written.

New options: `db_tags_table` (default `cache_tags`), `db_tags_col` (default `item_tag`), `db_tags_tag_index_name` (default `idx_cache_tags_item_tag`).

Documentation should cover: the second table and its setup via `createTable()`, the three new options, and when to pick this over `TagAwareAdapter` around `PdoAdapter`.

stof commented on src/Symfony/Component/Cache/Adapter/PdoTagAwareAdapter.php:
```diff
@@ -0,0 +1,239 @@
+            // - language processing like é == e
+            'mysql' => "CREATE TABLE $this->tagsTable ($this->idCol VARBINARY(255) NOT NULL, $this->tagCol VARBINARY(255) NOT NULL, PRIMARY KEY($this->idCol, $this->tagCol), INDEX idx_id_col ($this->idCol), INDEX idx_tag_col ($this->tagCol)) COLLATE utf8mb4_bin, ENGINE = InnoDB",
+            'sqlite' => "CREATE TABLE $this->tagsTable ($this->idCol TEXT NOT NULL, $this->tagCol TEXT NOT NULL, PRIMARY KEY ($this->idCol, $this->tagCol));CREATE INDEX idx_id_col ON $this->tagsTable($this->idCol);CREATE INDEX idx_tag_col ON $this->tagsTable($this->tagCol)",
+            'pgsql', 'sqlsrv' => "CREATE TABLE $this->tagsTable ($this->idCol VARCHAR(255) NOT NULL, $this->tagCol VARCHAR(255) NOT NULL, PRIMARY KEY($this->idCol, $this->tagCol);CREATE INDEX idx_id_col ON $this->tagsTable($this->idCol);CREATE INDEX idx_tag_col ON $this->tagsTable($this->tagCol)",
```
index names should be generated by including the `tagsTable` name IMO, as index names need to be unique at the level of the database schema, not at the level of the table.

stof commented on src/Symfony/Component/Cache/Adapter/PdoTagAwareAdapter.php:
```diff
@@ -0,0 +1,239 @@
+            // - language processing like é == e
+            'mysql' => "CREATE TABLE $this->tagsTable ($this->idCol VARBINARY(255) NOT NULL, $this->tagCol VARBINARY(255) NOT NULL, PRIMARY KEY($this->idCol, $this->tagCol), INDEX idx_id_col ($this->idCol), INDEX idx_tag_col ($this->tagCol)) COLLATE utf8mb4_bin, ENGINE = InnoDB",
+            'sqlite' => "CREATE TABLE $this->tagsTable ($this->idCol TEXT NOT NULL, $this->tagCol TEXT NOT NULL, PRIMARY KEY ($this->idCol, $this->tagCol));CREATE INDEX idx_id_col ON $this->tagsTable($this->idCol);CREATE INDEX idx_tag_col ON $this->tagsTable($this->tagCol)",
+            'pgsql', 'sqlsrv' => "CREATE TABLE $this->tagsTable ($this->idCol VARCHAR(255) NOT NULL, $this->tagCol VARCHAR(255) NOT NULL, PRIMARY KEY($this->idCol, $this->tagCol);CREATE INDEX idx_id_col ON $this->tagsTable($this->idCol);CREATE INDEX idx_tag_col ON $this->tagsTable($this->tagCol)",
```
The index on `idCol` is redundant as the primary key is already covering it. The extra index makes writes slower (as they need to update the index) without any benefits for reads (as there is already another usable index)

Toflar replied:
Done.

stof commented on src/Symfony/Component/Cache/Adapter/PdoTagAwareAdapter.php:
```diff
@@ -0,0 +1,239 @@
+class PdoTagAwareAdapter extends AbstractTagAwareAdapter implements PruneableInterface
+{
+    use PdoTrait {
+        doSave as public doSaveItem;
```
why public ?

Toflar replied:
Fixed.

stof commented on src/Symfony/Component/Cache/Adapter/PdoTagAwareAdapter.php:
```diff
@@ -0,0 +1,239 @@
+    {
+        $conn = $this->getConnection();
+
+        $stmt = $conn->prepare("DELETE FROM $this->tagsTable WHERE $this->idCol NOT IN (SELECT $this->idCol FROM $this->table)");
```
should this adapter use a foreign key with `ON DELETE CASCADE` instead ?

Toflar replied:
I decided against this on purpose. It's simpler like that 🤷‍♂️

stof replied:
why is it simpler to require handling orphaned tags explicitly ?

Toflar replied:
Sure, if you think that is better, please provide the necessary changes. I have no clue how `oci` and `sqlsrv` work and no motivation to learn about them 😄 We can also just do it for the others and throw an exception for those ones and ask the community to finish the implementation. Whoever needs it. But the current query likely works on all platforms.

stof commented on src/Symfony/Component/Cache/Adapter/PdoTagAwareAdapter.php:
```diff
@@ -0,0 +1,239 @@
+
+    protected function doDeleteTagRelations(array $tagData): bool
+    {
+        foreach ($tagData as $tagId => $idList) {
```
is there a chance `$idList` could be an empty array at the point ? This would break the query.

Toflar replied:
I don't think there is but I have added a fallback anyway.

stof commented on src/Symfony/Component/Cache/Traits/PdoTrait.php:
```diff
@@ -0,0 +1,409 @@
+    private string $dataCol = 'item_data';
+    private string $lifetimeCol = 'item_lifetime';
+    private string $timeCol = 'item_time';
+    private string $tagCol = 'item_tag';
```
This should not be in the trait but in the tag-aware adapter IMO.

Toflar replied:
Right. Fixed.

OskarStark commented on src/Symfony/Component/Cache/Adapter/PdoTagAwareAdapter.php:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    /**
+     * Creates the table to store cache items which can be called once for setup.
```
```suggestion
     * Creates the tables to store cache items which can be called once for setup.
```
?

OskarStark commented on src/Symfony/Component/Cache/Adapter/PdoTagAwareAdapter.php:
```diff
@@ -0,0 +1,248 @@
+    /**
+     * Creates the table to store cache items which can be called once for setup.
+     *
+     * Cache ID are saved in a column of maximum length 255. Cache data is
+     * saved in a BLOB.
```
```suggestion
     * Cache IDs are saved in a column of maximum length 255.
     * Cache data is saved in a BLOB.
```

OskarStark commented on src/Symfony/Component/Cache/Traits/PdoTrait.php:
```diff
@@ -0,0 +1,407 @@
+    /**
+     * Creates the table to store cache items which can be called once for setup.
+     *
+     * Cache ID are saved in a column of maximum length 255. Cache data is
```
```suggestion
     * Cache IDs are saved in a column of maximum length 255. Cache data is
```

stof commented on src/Symfony/Component/Cache/Adapter/PdoTagAwareAdapter.php:
```diff
@@ -0,0 +1,248 @@
+
+    private string $tagsTable = 'cache_tags';
+    private string $tagCol = 'item_tag';
+    private string $tagIdxName = 'idx_cache_tags_item_tag';
```
The constructor should probably support customizing those though options

Toflar replied:
Done.

Toflar commented:
Well I have incorporated the requested changes and then never heard back regarding the Psalm issue. I have rebased onto 7.3 now which is about all I can do I guess 😊 Failed 8.4 tests are unrelated.

dmitryuk commented:
Any plans to merge? Or may be @Toflar will close the PR?

Toflar commented:
Nothing has changed, I still think it's a valuable addition.

dmitryuk commented:
@Toflar can you resolve conflicts and fix tests?

Toflar committed:
Add PdoTagAwareAdapter

Toflar committed:
CS

Toflar committed:
[Cache] Add PdoTagAwareAdapter

Tags are stored in their own table, one (item, tag) row each, so saving,
invalidating and pruning are plain relational statements instead of the
extra round trips TagAwareAdapter needs on top of PdoAdapter. The shared
PDO plumbing moves from PdoAdapter into a new PdoTrait, and TAGS_PREFIX
becomes overridable so the adapter can store bare tag names: with a
dedicated table there is no shared key space to disambiguate.
